### PR TITLE
Add ability to modify user name and email address associated with account

### PR DESCRIPTION
**Pull request by kellymclaughlin**
This has been requested by multiple customers now and would be generally useful.

### CODE REVIEW

**Comment by reiddraper** [comment]
A `src/#riak_cs_wm_usage.erl#` file snuck in there. Not sure what that is.

File: src/riak_cs_utils.erl [hunk 1]
```diff
@@ -734,9 +735,14 @@ riak_connection(Pool) ->
...
-    %% are siblings all copies should be the same
-    UpdUserObj = update_obj_value(UserObj, riak_cs_utils:encode_term(User)),
+    Indexes = [{?EMAIL_INDEX, User?RCS_USER.email},
+               {?ID_INDEX, User?RCS_USER.canonical_id}],
+    MD = dict:store(?MD_INDEX, Indexes, dict:new()),
+    UpdUserObj = riakc_obj:update_metadata(
+                   riakc_obj:update_value(UserObj,
+                                          riak_cs_utils:encode_term(User)),
+                   MD),
+    %% UpdUserObj = update_obj_value(UserObj, riak_cs_utils:encode_term(User)),
```
Comment (reiddraper) [inline src/riak_cs_utils.erl]: can remove this line

File: src/riak_cs_wm_user.erl [hunk 2]
```diff
@@ -107,25 +107,16 @@ create_path(RD, Ctx) -> {"/riak-cs/user", RD, Ctx}.
...
-                    riak_cs_s3_response:api_error(Reason, RD, Ctx)
-            end;
-        {'EXIT', _} ->
-            riak_cs_s3_response:api_error(invalid_user_update, RD, Ctx)
-    end;
+    Body = riak_cs_json:from_json(wrq:req_body(RD)),
+    {UserName, Email} =
+        riak_cs_json:value_or_default(
+          riak_cs_json:get(Body, [{<<"name">>, <<"email">>}]),
+          {<<>>, <<>>}),
```
Comment (reiddraper) [inline src/riak_cs_wm_user.erl:118]: Am I correct that this would allow empty `name` and `email` fields? Do we want that?

**Comment by reiddraper** [comment]
Am I correct that we already had the functionality to update the access secret?

**Comment by kellymclaughlin** [comment]
> Am I correct that we already had the functionality to update the access secret?

That is correct

File: src/riak_cs_wm_user.erl (hunk 2 above)
Comment (kellymclaughlin) [inline src/riak_cs_wm_user.erl:118]: You could potentially have an empty `name` field, but an empty email field would be disallowed by [`validate_email`](https://github.com/basho/riak_cs/blob/develop/src/riak_cs_utils.erl#L1313).

**Comment by reiddraper** [comment]
- [x] Eunit and eqc tests pass
- [x] Pulse tests
- [x] Client and int tests pass
- [x] Dialyzer output clean
- [x] Xref output clean
- [x] `riak_test` tests pass

**Comment by reiddraper** [comment]
I see two dialyzer `riak_cs_wm_user` warnings:

```
riak_cs_wm_user.erl:113: The call riak_cs_json:get(Body::[any()] | {'error','decode_failed'} | {'struct',_},[{<<_:32>>,<<_:40>>},...]) breaks the contract ({'struct',term()} | [term()] | 'undefined',path()) -> term()
riak_cs_wm_user.erl:390: Invalid type specification for function riak_cs_wm_user:new_user_response/4. The success typing is ({'error',atom() | {'error',atom() | {'error',atom() | {_,_}} | {'riak_connect_failed',_}} | {'riak_connect_failed',_}} | {'ok',#rcs_user_v2{name::string(),display_name::string(),email::string(),key_id::string(),key_secret::string(),canonical_id::string(),buckets::[],status::'enabled'}},[47 | 97 | 99 | 105 | 108 | 109 | 110 | 111 | 112 | 116 | 120,...],#wm_reqdata{wm_state::{'wm_reqstate',_,_,_,_,_,_,_,_},resp_headers::gb_tree()},#context{start_time::'undefined' | {non_neg_integer(),non_neg_integer(),non_neg_integer()},auth_bypass::atom(),user_object::'undefined' | {'riakc_obj','undefined' | binary(),'undefined' | binary(),'undefined' | binary(),'undefined' | [any()],'undefined' | dict(),'undefined' | binary()},bucket::'undefined' | binary(),requested_perm::'FULL_CONTROL' | 'READ' | 'READ_ACP' | 'WRITE' | 'WRITE_ACP' | 'undefined',riakc_pid::'undefined' | pid(),riakc_pool::atom(),submodule::atom(),exports_fun::'undefined' | fun(),auth_module::atom(),response_module::atom(),policy_module::atom(),api::atom()}) -> {'true' | {'halt',_},#wm_reqdata{},_}
```

**Comment by kellymclaughlin** [comment]
Ready for more review now. I updated `user_test` to do a much more thorough job of exercising the user API including testing user listing and new `key_secret` issuance. Also found the bug that caused `500` to be returned on user requests that failed authentication checking instead of `403`.

**Comment by shino** [comment]
Sorry for breaking in but I have a question.
CS user's email address should be unique in system wide https://github.com/basho/riak_cs/wiki/Globally-Unique-Entities . This PR does not depend on / include Stanchion change. Should we forbid email duplication?

**Comment by kellymclaughlin** [comment]
> Sorry for breaking in but I have a question.
> CS user's email address should be unique in system wide https://github.com/basho/riak_cs/wiki/Globally-Unique-Entities . This PR does not depend on / include Stanchion change. Should we forbid email duplication?

@shino Yep, you are exactly right. I'm going to work on this today. Thanks for pointing that out.

**Comment by kellymclaughlin** [comment]
I've gotten serialized user updates working and should have PRs for `riak_cs`, `stanchion`, and `velvet` in tomorrow, but I think it should be handled and reviewed separately from these changes.

I've open #611 for to track this work.

**Comment by reiddraper** [comment]
+1

**Pull request closed without merging**